Validate if a Pipeline contains a cycle

EstebanFS opened the pull request:
<!--
Thanks for submitting a pull request to Concourse!

If you haven't already, feel free to [add yourself] as a contributor so that
you can add labels to your PR and re-trigger its builds if they fail.

Also check the [PR requirements] if you haven't already!
-->

[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR

Now we can detect if the pipelines contains a cycle before creating it.

<!--
Summarize your changes as a checklist, leaving any unfinished work as unchecked
items. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->

* [x] Adding cycle detection on the validate pipeline method

## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->

if you execute `fly set-pipeline` with the following pipeline that contains a cycle

```yml
resources:
- name: get_1
  type: mock

jobs:
- name: step_1
  plan:
  - in_parallel:
      steps:
      - get: get_1
        passed:
        - step_1
```

you should see the following error `pipeline contains a cycle that starts at Job 'step_1'`

## Release Note

* The API will reject any pipelines that contains a cycle

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

taylorsilva committed:
Validate if a Pipeline contains a cycle

Signed-off-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Taylor Silva  <tasilva@vmware.com>

aoldershaw commented on atc/configvalidate/validate.go:
```diff
@@ -490,3 +496,62 @@ func validateDisplay(c atc.Config) ([]atc.ConfigWarning, error) {
 	return warnings, nil
 }
+
+func detectCycle(j atc.JobConfig, visited map[string]int, initialJobConfig atc.Config) (bool, string) {
```
Maybe rename `initialJobConfig` to `pipelineConfig`

aoldershaw commented on atc/configvalidate/validate.go:
```diff
@@ -490,3 +496,62 @@ func validateDisplay(c atc.Config) ([]atc.ConfigWarning, error) {
+				if visited[nextJobName] == semiVisited {
+					state = true
+					cycleErrorJob = nextJobName
+					return nil
```
Not a big deal, but WDYT about changing the signature of `detectCycle` to just return an `error`. If we detect a cycle here, we can just return the error we currently construct here: https://github.com/concourse/concourse/blob/c948550fecc2a8df7fd178364aed7d145b15cc80/atc/configvalidate/validate.go#L554

Then, we can simply return the `error` that's returned by `j.StepConfig.Visit(...)`, and don't need to keep track of `state` or `cycleErrorJob`, simplifying the control flow a bit. Plus, it wasn't totally obvious to me what `state` referred to initially - but returning an `error` makes it more obvious that this is where the cycle occurs

aoldershaw commented on atc/configvalidate/validate.go:
```diff
@@ -490,3 +496,62 @@ func validateDisplay(c atc.Config) ([]atc.ConfigWarning, error) {
+	for _, job := range jobs {
+		cycleExist, cycleErrorJob = detectCycle(job, emptyMap, c)
+		if cycleExist == true {
+			break
```
Similarly here, maybe we can just return the error here rather than `break`ing, so we can scope `cycleExist` and `cycleErrorJob` to within the `for` loop (or if you decide to return an `error` from `detectCycle` instead, this just becomes `if err != nil { return err }`)

aoldershaw reviewed (changes requested):
Had to sketch it out to understand how it works, but it's a neat algorithm!

I left some suggestions for making the control a bit simpler to follow, at least IMO. Let me know if you have any questions about it, and feel free to push back if you disagree

taylorsilva committed:
Added code improvements

Simplifying control flow on function returnings
Changing some variables names to be more readable

Signed-off-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Taylor Silva  <tasilva@vmware.com>

aoldershaw reviewed (approved):
LGTM! Thanks for making the changes I requested
